casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -86,13 +86,54 @@ object EstimatorHelper {
       ancestors = maybeAncestors.flatten
       ancestorEvents = ancestors.flatMap(_.getBody.deploys.flatMap(_.deployLog)) ++
         ancestors.flatMap(_.getBody.deploys.flatMap(_.paymentLog))
+      (produceEvents, consumeEvents, commEvents) = ancestorEvents.foldLeft(
+        (List.empty[ProduceEvent], List.empty[ConsumeEvent], List.empty[CommEvent])
+      ) {
+        case ((produces, consumes, commEvents), ancestorEvent) =>
+          ancestorEvent match {
+            case Event(Produce(produce: ProduceEvent)) =>
+              (produces :+ produce, consumes, commEvents)
+            case Event(Consume(consume: ConsumeEvent)) =>
+              (produces, consumes :+ consume, commEvents)
+            case Event(Comm(commEvent: CommEvent)) =>
+              (produces, consumes, commEvents :+ commEvent)
+            case _ =>
+              throw new RuntimeException("Unexpected ancestor event")
+          }
+      }
+      producesInCommEvents = commEvents.flatMap {
+        case CommEvent(Some(_: ConsumeEvent), produces) =>
+          produces
+        case _ =>
+          throw new RuntimeException("Unexpected comm event")
```

Comment:
I don't think this case is needed here, as we have a list of CommEvents here - the type system guards this, and the totality checker should see that. There must be a way to have a total function commEvents -> producesInCommEvents
...
Ah, I see... Damn protobuf and its option-ridden garbage output.

The error should then say `Invalid comm event $c`, where `c` is part of the pattern in the first case `c @ CommEvent(...)`